.Lib9c.Tools/SubCommand/Market.cs

```diff
@@ -0,0 +1,189 @@
+                $"#IDX".PadRight(indexWidth),
+                "HASH".PadRight(BlockHash.Size * 2),
+                "TIME".PadRight(DateTimeOffset.Now.ToString("o").Length),
+                "TXID".PadRight(TxId.Size * 2),
+                "BUYER".PadRight(Address.Size * 2),
+                "BUYER AVATER".PadRight(Address.Size * 2),
+                "SELLER".PadRight(Address.Size * 2),
+                "SELLER AVATER".PadRight(Address.Size * 2),
+                "PRICE"
+            );
+
+            while (true)
+            {
+                stderr.WriteLine("Scanning block #{0} {1}...", block.Index, block.Hash);
+                stderr.Flush();
+
+                foreach (var tx in block.Transactions.Reverse())
+                if (includeFails ||
+                    !(chain.GetTxExecution(block.Hash, tx.Id) is { } e) ||
+                    e is TxSuccess)
+                foreach (var act in tx.Actions.Reverse())
+                {
+                    ActionBase a = act.InnerAction;
+                    IEnumerable<Order> orders = act.InnerAction switch
+                    {
+                        IBuy0 b0 => new Order[]
+                        {
+                            new Order
+                            {
+                                BuyerAvatar = b0.buyerAvatarAddress,
+                                Seller = b0.sellerAgentAddress,
+                                SellerAvatar = b0.sellerAvatarAddress,
+                            },
+                        },
+                        IBuy5 b => b.purchaseInfos.Reverse().Select(p => new Order
+                        {
+                            BuyerAvatar = b.buyerAvatarAddress,
+                            Seller = p.SellerAgentAddress,
+                            SellerAvatar = p.SellerAvatarAddress,
+                            Price = p.Price,
+                            ItemSubType = p.ItemSubType,
+                        }),
+                        _ => new Order[0],
+                    };
+
+                    if (itemTypes is {} types)
+                    {
+                        orders = orders.Where(o => o.ItemSubType is {} t && types.Contains(t));
+                    }
+
+                    foreach (Order order in orders)
+                    {
+                        Console.WriteLine(
+                            "{0}\t{1}\t{2}\t{3}\t{4}\t{5}\t{6}\t{7}\t{8}",
+                            $"#{block.Index}".PadLeft(indexWidth),
+                            block.Hash,
+                            tx.Timestamp.ToString("o"),
+                            tx.Id,
+                            tx.Signer.ToHex(),
+                            order.BuyerAvatar.ToHex(),
+                            order.Seller.ToHex(),
+                            order.SellerAvatar.ToHex(),
+                            order.Price?.ToString() ?? "(N/A)"
+                        );
+                    }
+                }
```

Comment:
How about wrapping this lines with braces? (Of course then this will be with very deep indent 😥....)

Reply:
Or how about separating these lines to private method to receive block? These lines seem like print orders from each block.